### PR TITLE
Port chain events over to event.Feed

### PR DESCRIPTION
**Pull request by tailingchen**
This PR ports core.TxPreEvent over to event.Feed according to the comments in #14840 . 
@fjl  Please help to review this PR, thanks.

### CODE REVIEW

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 99a55a862299457705066a53b7b8a95411f201ce
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: miner/worker.go [hunk 1]
```diff
@@ -138,7 +154,11 @@ func newWorker(config *params.ChainConfig, engine consensus.Engine, coinbase com
 		unconfirmed:    newUnconfirmedBlocks(eth.BlockChain(), 5),
 		fullValidation: false,
 	}
-	worker.events = worker.mux.Subscribe(core.ChainHeadEvent{}, core.ChainSideEvent{}, core.TxPreEvent{})
+	// Subscribe TxPreEvent for tx pool
+	worker.txSub = eth.TxPool().SubscribeTxPreEvent(worker.txCh)
+	// Subscribe events for blockchain
+	worker.chainHeadSub = eth.BlockChain().SubscribeChainHeadEvent(worker.chainHeadCh)
+	worker.chainSideSub = eth.BlockChain().SubscribeChainSideEvent(worker.chainSideCh)
```
Comment (fjl) [inline miner/worker.go]: `Unsubscribe` needs to be called somewhere. You can just add a few `defer worker.x.Unsubscribe()` calls to `update`.

File: miner/worker.go [hunk 2]
```diff
@@ -303,12 +335,18 @@ func (self *worker) wait() {
...
+					events = append(events, core.ChainEvent{Block: block, Hash: block.Hash(), Logs: logs})
 
 					if stat == core.CanonStatTy {
-						self.mux.Post(core.ChainHeadEvent{Block: block})
-						self.mux.Post(logs)
+						events = append(events, core.ChainHeadEvent{Block: block})
+						coalescedLogs = logs
 					}
+					// post blockchain events
+					self.chain.PostChainEvents(events, coalescedLogs)
```
Comment (fjl) [inline miner/worker.go]: Not super happy about exposing `postChainEvents`, but I understand it's the quickest solution for now. I'll try to move events into WriteBlock later.

File: miner/worker.go (hunk 1 above)
Comment (tailingchen) [inline miner/worker.go]: Done. Fixed in new version.

File: miner/worker.go (hunk 2 above)
Comment (tailingchen) [inline miner/worker.go]: Sure. I add TODO comments to figure out it.

File: core/blockchain.go [hunk 3]
```diff
@@ -1184,39 +1197,42 @@ func (bc *BlockChain) reorg(oldBlock, newBlock *types.Block) error {
...
-			// even necessary.
-			if bc.LastBlockHash() == event.Hash {
-				bc.eventMux.Post(ChainHeadEvent{event.Block})
-			}
+		switch ev := event.(type) {
+		case ChainEvent:
+			bc.chainFeed.Send(ev)
+
+		case ChainHeadEvent:
+			bc.chainHeadFeed.Send(ChainHeadEvent{ev.Block})
```
Comment (bas-vk) [inline core/blockchain.go]: Any reason why you wrap the block in a new `ChainHeadEvent`?

Comment (tailingchen) [inline core/blockchain.go]: It's my bad. I forgot to modify it. Already fixed in new version. Thanks for reviewing.

**Comment by tailingchen** [comment]
@fjl fixed. Help to quick review.

File: eth/filters/filter_system.go [hunk 4]
```diff
@@ -395,7 +401,19 @@ func (es *EventSystem) lightFilterLogs(header *types.Header, addresses []common.
...
+		// Subscribe RemovedLogsEvent
+		rmLogsCh  = make(chan core.RemovedLogsEvent, rmLogsChanSize)
+		rmLogsSub = es.backend.SubscribeRemovedLogsEvent(rmLogsCh)
+		// Subscribe []*types.Log
+		logsCh  = make(chan []*types.Log, logsChanSize)
+		logsSub = es.backend.SubscribeLogsEvent(logsCh)
+		// Subscribe ChainEvent
+		chainEvCh  = make(chan core.ChainEvent, chainEvChanSize)
+		chainEvSub = es.backend.SubscribeChainEvent(chainEvCh)
 	)
```
Comment (fjl) [inline eth/filters/filter_system.go]: Unsubscribe is missing here too. Sorry, I missed that earlier.

Comment (tailingchen) [inline eth/filters/filter_system.go]: @fjl Fixed in new version. Thanks!

**Comment by fjl** [comment]
Let's see what breaks ;)

**Pull request merged**